README.md
# File Uploader by Uploadcare

<a href="https://uploadcare.com/?utm_source=github&utm_campaign=uploadcare-tinymce">
  <img align="right" width="64" height="64"
    src="https://ucarecdn.com/2f4864b7-ed0e-4411-965b-8148623aa680/uploadcare-logo-mark.svg"
    alt="">
</a>

This is a plugin for [TinyMCE][tinymce], a WYSIWYG HTML editor,
providing it for working with [Uploadcare Widget][uc-feature-widget].

The plugin allows TinyMCE users to upload media
from their devices, social media, cloud storage, and more.
All that without any backend code that's usually required to handle uploads.

[![GitHub release][badge-release-img]][badge-release-url]&nbsp;
[![Uploadcare stack on StackShare][badge-stack-img]][badge-stack-url]

* [Demo](#demo)
* [Requirements](#requirements)
* [Install](#install)
* [Usage](#usage)
* [Configuration](#configuration)
  * [Plugin configuration](#plugin-configuration)
  * [Widget configuration](#widget-configuration)
* [Security issues](#security-issues)
* [Feedback](#feedback)

## Demo

Check out the basic demo for TinyMCE 4.x [here][demo-4].

And for TinyMCE 5.x — [here][demo-5].

## Requirements

TinyMCE 4.x or TinyMCE 5.x

## Install

Download the latest plugin archive from the [release branch][github-branch-release]
or [releases page][github-releases].

Extract the downloaded archive to the plugin directory of your TinyMCE
installation.

Another option here is cloning the repo:

```bash
git clone -b release git@github.com:uploadcare/uploadcare-tinymce.git plugins/uploadcare
```

## Usage

Add `uploadcare` to the list of your TinyMCE plugins and the toolbar.
**Set your [public key][uc-docs-widget-options-public-key]**. Public keys are used
to identify a target Uploadcare [project][uc-projects] your uploads will go to.

```javascript
tinymce.init({
  selector: '#editor',
  plugins: 'uploadcare',
  toolbar: 'uploadcare',
  external_plugins: {
    uploadcare: '/path/to/uploadcare/plugin.js',
  },
  uploadcare_public_key: 'YOUR_PUBLIC_KEY',
})
```

## Configuration

### Plugin configuration

To apply a custom configuration, initialize the TinyMCE editor
providing additional options:

```javascript
tinymce.init({
  selector: '#editor',
  plugins: 'uploadcare',
  toolbar: 'uploadcare',
  external_plugins: {
    uploadcare: '/path/to/uploadcare/plugin.js',
  },
  uploadcare_public_key: 'YOUR_PUBLIC_KEY',
  /* when handling images, you can resize them on a client to save bandwidth */
  uploadcare_image_shrink: '500x375',
  /* allow multi-file uploads */
  uploadcare_multiple: true,
  uploadcare_multiple_max: 3,
  /* set crop options when handling images */
  uploadcare_crop: '1:1,4:3',
  /* feel free to add more options here */
})
```

You can use any [widget options][uc-docs-widget-options]
providing those as object keys. Just add the `uploadcare_` prefix and
use `snake_case` instead of `camelCase` in option names, e.g.
`imagesOnly` &rarr; `uploadcare_images_only`.

### Widget configuration

Uploadcare Widget can be deeply customized to suit your UX/UI. You can define
allowed upload sources, implement file validation, and more.

Use our live [widget sandbox][uc-widget-configure] as a starting point and consider
checking out the docs on [widget configuration][uc-docs-widget-config] and its
[JavaScript API][uc-docs-widget-js-api].

## Security issues

If you think you ran into something in Uploadcare libraries which might have
security implications, please hit us up at [bugbounty@uploadcare.com][uc-email-bounty]
or Hackerone.

We'll contact you personally in a short time to fix an issue through co-op and
prior to any public disclosure.

## Feedback

Issues and PRs are welcome. You can provide your feedback or drop us a support
request at [hello@uploadcare.com][uc-email-hello].

[tinymce]: http://www.tinymce.com/
[demo-4]: https://uploadcare.github.io/uploadcare-tinymce/demo/4.x/?utm_source=github&utm_campaign=uploadcare-tinymce
[demo-5]: https://uploadcare.github.io/uploadcare-tinymce/demo/5.x/?utm_source=github&utm_campaign=uploadcare-tinymce
[uc-docs-widget-config]: https://uploadcare.com/docs/uploads/widget/config/?utm_source=github&utm_campaign=uploadcare-tinymce
[uc-docs-widget-js-api]: https://uploadcare.com/docs/api_reference/javascript/?utm_source=github&utm_campaign=uploadcare-tinymce
[uc-docs-widget-options]: https://uploadcare.com/docs/uploads/widget/config/?utm_source=github&utm_campaign=uploadcare-tinymce#options
[uc-docs-widget-options-public-key]: https://uploadcare.com/docs/uploads/widget/config/?utm_source=github&utm_campaign=uploadcare-tinymce#option-public-key
[uc-dashboard]: https://uploadcare.com/dashboard/?utm_source=github&utm_campaign=uploadcare-tinymce
[uc-widget-configure]: https://uploadcare.com/widget/configure/?utm_source=github&utm_campaign=uploadcare-tinymce
[uc-feature-widget]: https://uploadcare.com/features/widget/?utm_source=github&utm_campaign=uploadcare-tinymce
[uc-projects]: https://uploadcare.com/docs/keys/?utm_source=github&utm_campaign=uploadcare-tinymce#projects
[uc-email-bounty]: mailto:bugbounty@uploadcare.com
[uc-email-hello]: mailto:hello@uploadcare.com
[github-releases]: https://github.com/uploadcare/uploadcare-tinymce/releases
[github-branch-release]: https://github.com/uploadcare/uploadcare-tinymce/tree/release
[github-contributors]: https://github.com/uploadcare/uploadcare-tinymce/graphs/contributors
[badge-stack-img]: https://img.shields.io/badge/tech-stack-0690fa.svg?style=flat
[badge-stack-url]: https://stackshare.io/uploadcare/stacks/
[badge-release-img]: https://img.shields.io/github/release/uploadcare/uploadcare-tinymce.svg
[badge-release-url]: https://github.com/uploadcare/uploadcare-tinymce/releases


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

[Unreleased]: https://github.com/uploadcare/uploadcare-tinymce/compare/v3.2.2...HEAD

## [3.2.2], 2019-06-17

### Fixed

* Fix dist tarball generation due to outdated `release-it` config.

## [3.2.1], 2019-06-17

### Fixed

* Fix missing png icon for TinyMCE 4.x.

## [3.2.0], 2019-06-17

### Changed

* Add support for TinyMCE 5.x. See [#11][github-pr-11].

[github-pr-11]: https://github.com/uploadcare/uploadcare-tinymce/pull/11

## [3.1.1], 2019-06-17

### Changed

* Update button title from `Uploadcare` to `Insert media` to make it clearer.
  See [#12][github-pr-12].

[github-pr-12]: https://github.com/uploadcare/uploadcare-tinymce/pull/12

## [3.1.0], 2019-02-21

### Added

* The ability to edit the inserted images
  and links to uploaded files. See [#9][github-pr-9].

[3.1.0]: https://github.com/uploadcare/uploadcare-tinymce/compare/v3.0.0...v3.1.0
[github-pr-9]: https://github.com/uploadcare/uploadcare-tinymce/pull/9

## 3.0.0, 2018-05-16

### Added

* Plugin settings can now be defined using the TinyMCE editor options.

### Changed

* Bumped the widget version to `3.x`, see [widget changelog][widget-changelog].
* Added `integration` to the default config to provide info
  on the TinyMCE version and plugin version for User Agent with requests
  to [Uploadcare Upload API](https://uploadcare.com/docs/api_reference/upload/).

### Removed

* Setting options for the plugin using the `config.js` file.

## 2.1.1, 2017-01-16

### Changed

* CDN link to the widget.

## 2.1.0, 2016-11-30

### Changed

* Bumped widget version to `2.10.2`, see [widget changelog][widget-changelog].

## 2.0.0, 2015-09-08

### Changed

* Updated plugin to work with TinyMCE 4+.
* Get rid of PHP dialog.
* Bumped widget version to 2.5.1, see [widget changelog][widget-changelog].

[widget-changelog]: https://github.com/uploadcare/uploadcare-widget/blob/master/HISTORY.markdown


plugin.js
/**
 * uploadcare-tinymce 3.2.2
 * File Uploader by Uploadcare, a plugin providing TinyMCE users to upload media via Uploadcare Widget.
 * https://github.com/uploadcare/uploadcare-tinymce#readme
 * Date: 2019-06-17
 */

(function () {
  'use strict';

  var icon = "\n<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"32\" height=\"32\" viewBox=\"0 0 80 80\">\n  <path fill=\"#FFD800\" fill-rule=\"evenodd\" d=\"M32,64 C14.326888,64 0,49.673112 0,32 C0,14.326888 14.326888,0 32,0 C49.673112,0 64,14.326888 64,32 C64,49.673112 49.673112,64 32,64 Z M32,34.7654321 C33.527306,34.7654321 34.7654321,33.527306 34.7654321,32 C34.7654321,30.472694 33.527306,29.2345679 32,29.2345679 C30.472694,29.2345679 29.2345679,30.472694 29.2345679,32 C29.2345679,33.527306 30.472694,34.7654321 32,34.7654321 Z\" transform=\"translate(8 8)\"/>\n</svg>\n";

  /* global tinymce, uploadcare */

  var uploadcareDefaultOptions = {
    integration: getIntegration(),
    crop: '',
  };

  tinymce.create('tinymce.plugins.UploadcarePlugin', {
    init: function(editor, url) {
      tinymce.ScriptLoader.add('https://ucarecdn.com/widget/' + '3.x' + '/uploadcare/uploadcare.full.min.js');
      tinymce.ScriptLoader.loadQueue();

      var uploadcareOptions = Object.keys(editor.settings)
        .filter(function(settingName) {
          return settingName.search('^uploadcare_') !== -1
        })
        .reduce(function(options, settingName) {
          var optionName = settingName
            .replace('uploadcare_', '')
            .replace(/(_[a-z])/g, function(v) {
              return v.toUpperCase()
            })
            .replace('_', '');

          options[optionName] = editor.settings[settingName];

          return options
        }, uploadcareDefaultOptions);

      function showUploadcareDialog() {
        var cdnBase = uploadcareOptions.cdnBase || uploadcare.defaults.cdnBase;
        var file = null;
        var selectedNode = editor.selection.getNode();

        if (selectedNode.nodeName === 'IMG') {
          if (selectedNode.src.indexOf(cdnBase) === 0) {
            file = uploadcare.fileFrom('uploaded', selectedNode.src, uploadcareOptions);
          }
        }
        if (selectedNode.nodeName === 'A') {
          if (selectedNode.href.indexOf(cdnBase) === 0) {
            file = uploadcare.fileFrom('uploaded', selectedNode.href, uploadcareOptions);
          }
        }

        uploadcare.openDialog(file, uploadcareOptions).done(function(file) {
          file.done(function(fileInfo) {
            if (fileInfo.isImage) {
              editor.selection.setNode(tinymce.activeEditor.dom.create('img', {src: fileInfo.cdnUrl}));
            }
            else if (selectedNode.nodeName === 'A') {
              selectedNode.parentNode.replaceChild(
                tinymce.activeEditor.dom.create('a', {href: fileInfo.cdnUrl}, fileInfo.name),
                selectedNode
              );
            }
            else {
              editor.selection.setNode(tinymce.activeEditor.dom.create('a', {href: fileInfo.cdnUrl}, fileInfo.name));
            }
          });
        });
      }

      if (detectMajorVersion() === 4) {
        editor.addCommand('showUploadcareDialog', showUploadcareDialog);

        editor.addButton('uploadcare', {
          title: 'Insert media',
          cmd: 'showUploadcareDialog',
          image: url + '/icons/uploadcare.png',
          stateSelector: 'img',
        });
      }
      else if (detectMajorVersion() === 5) {
        editor.ui.registry.addIcon('uploadcare', icon);

        editor.ui.registry.addButton('uploadcare', {
          text: 'Insert media',
          tooltip: 'Insert media',
          onAction: showUploadcareDialog,
          icon: 'uploadcare',
        });
      }
    },

    createControl: function() {
      return null
    },

    getInfo: function() {
      return {
        longname: 'Uploadcare',
        author: 'Uploadcare',
        authorurl: 'https://uploadcare.com/',
        infourl: 'https://github.com/uploadcare/uploadcare-tinymce',
        version: '3.2.2',
      }
    },
  });

  tinymce.PluginManager.add('uploadcare', tinymce.plugins.UploadcarePlugin);

  function getIntegration() {
    var tinymceVersion = tinymce.majorVersion + '.' + tinymce.minorVersion;
    var pluginVerion = '3.2.2';

    return 'TinyMCE/{tinymceVersion}; Uploadcare-TinyMCE/{pluginVerion}'
      .replace('{tinymceVersion}', tinymceVersion)
      .replace('{pluginVerion}', pluginVerion)
  }

  function detectMajorVersion() {
    var version = parseInt(tinymce.majorVersion);

    return version
  }

}());
